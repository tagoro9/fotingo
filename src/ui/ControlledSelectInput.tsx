import React = require('react');

import { pointer } from 'figures';
import { Box, Color } from 'ink';

interface ControlledSelectInputProps {
  items: Array<{
    key?: React.Key;
    label: string;
    value: React.Key;
  }>;
  selectedIndex?: number;
}

interface IndicatorProps {
  isSelected: boolean;
}

interface ItemProps {
  isSelected: boolean;
  label: string;
}

function Indicator({ isSelected }: IndicatorProps): JSX.Element {
  return <Box marginRight={1}>{isSelected ? <Color blue>{pointer}</Color> : ' '}</Box>;
}

function Item({ isSelected, label }: ItemProps): JSX.Element {
  return <Color blue={isSelected}>{label}</Color>;
}

export function ControlledSelectInput({
  items,
  selectedIndex,
}: ControlledSelectInputProps): JSX.Element {
  return (
    <Box flexDirection="column">
      {items.map((item, index) => {
        const isSelected = index === selectedIndex;

        return (
          <Box key={item.key || item.value}>
            <Indicator isSelected={isSelected} />
            <Item {...item} isSelected={isSelected} />
          </Box>
        );
      })}
    </Box>
  );
}
