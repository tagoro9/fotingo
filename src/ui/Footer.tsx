import * as humanize from 'humanize-duration';
import { Box } from 'ink';
import React = require('react');

interface FooterProps {
  executionTime: number;
}

export function Footer({ executionTime }: FooterProps) {
  return (
    <Box marginTop={1}>
      <Box marginRight={2}>âœ¨</Box>
      <Box>Done in {humanize(executionTime, { round: true })}.</Box>
    </Box>
  );
}
