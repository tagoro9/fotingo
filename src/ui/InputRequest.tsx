import React = require('react');

import { RequestProps, SelectRequestProps } from './props';
import { SelectRequest } from './SelectRequest';
import { TextRequest } from './TextRequest';

function isSelectRequest(props: RequestProps): props is SelectRequestProps {
  return 'options' in props.request;
}

export function InputRequest(props: RequestProps): JSX.Element {
  if (isSelectRequest(props)) {
    return <SelectRequest {...props} />;
  }
  return <TextRequest {...props} />;
}
