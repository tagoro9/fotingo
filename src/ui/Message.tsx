import React = require('react');

import { Box, Color } from 'ink';
import Spinner from 'ink-spinner';
import { get as getEmoji } from 'node-emoji';
import { MessageProps } from './props';

const MESSAGE_TYPE_TO_EMOJI: Record<MessageProps['message']['type'], string | undefined> = {
  error: '💥',
  info: '📝',
  request: '❔',
  status: undefined,
};

export function Message({ message, isDone = false, isLast = false }: MessageProps) {
  return (
    <Box>
      <Box marginRight={2}>
        {!isDone && message.showSpinner && isLast ? (
          <Color cyan>
            <Spinner type="dots" />
          </Color>
        ) : (
          (message.emoji && getEmoji(message.emoji)) || MESSAGE_TYPE_TO_EMOJI[message.type]
        )}
      </Box>
      <Box>
        <Color red={message.type === 'error'}>{message.message}</Color>
        {message.detail && (
          <Box marginLeft={1}>
            <Color grey={message.type === 'request'}>{message.detail}</Color>
          </Box>
        )}
      </Box>
    </Box>
  );
}
