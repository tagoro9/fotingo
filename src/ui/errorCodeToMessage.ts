import { GitErrorType } from 'src/git/GitError';

export const ERROR_CODE_TO_MESSAGE: { [k: number]: string } = {
  [GitErrorType.BRANCH_ALREADY_EXISTS]: 'It looks like there is a branch already for this issue',
  [GitErrorType.NOT_A_GIT_REPO]: 'It looks like you run fotingo outside of a git repository',
};
