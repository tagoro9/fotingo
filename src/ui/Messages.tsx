import React = require('react');

import { Box, Static } from 'ink';
import { init, last } from 'ramda';
import { Message } from './Message';
import { MessagesProps } from './props';

export function Messages({
  isDebugging,
  isDone,
  isInThread,
  isRequesting,
  messages,
}: MessagesProps) {
  const pastMessages = init(messages);
  const lastMessage = last(messages);
  const staticMessages =
    isDebugging && lastMessage !== undefined ? [...pastMessages, lastMessage] : pastMessages;
  return (
    <Box flexDirection="column">
      <Static>
        {staticMessages.map((msg, id) => (
          <Message key={id} message={msg} />
        ))}
      </Static>
      {lastMessage && !isDebugging && !isRequesting && !isInThread && (
        <Message isDone={isDone} isLast={true} message={lastMessage} />
      )}
    </Box>
  );
}
