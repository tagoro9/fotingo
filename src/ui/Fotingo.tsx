import React = require('react');

import { Footer } from './Footer';
import { InputRequest } from './InputRequest';
import { Messages } from './Messages';
import { FotingoProps } from './props';
import { useCmd } from './useCmd';

export function Fotingo({ cmd, messenger }: FotingoProps) {
  const { executionTime, isDone, isInThread, messages, request, sendRequestData } = useCmd(
    messenger,
    cmd,
  );

  return (
    <>
      <Messages
        isDone={isDone}
        isRequesting={request !== undefined}
        isInThread={isInThread}
        messages={messages}
      />
      {request && <InputRequest request={request} onSubmit={sendRequestData} />}
      {!request && executionTime && <Footer executionTime={executionTime} />}
    </>
  );
}
