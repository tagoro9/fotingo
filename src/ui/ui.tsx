import React = require('react');

import { render } from 'ink';

import { Fotingo } from './Fotingo';
import { FotingoProps } from './props';

export function renderUi(props: FotingoProps): void {
  render(<Fotingo {...props} />);
}
