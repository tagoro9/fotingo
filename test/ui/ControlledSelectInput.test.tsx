import 'jest';
import React = require('react');

import { render } from 'ink-testing-library';
import { ControlledSelectInput } from 'src/ui/ControlledSelectInput';

describe('<ControlledSelectInput />', () => {
  const items = [
    { label: 'Item 1', value: 1 },
    { label: 'Item 2', value: 2 },
  ];

  test('renders the list of items', () => {
    const props = {
      items,
    };
    const { lastFrame } = render(<ControlledSelectInput {...props} />);
    expect(lastFrame()).toMatchSnapshot();
  });

  test('renders the selected item with a pointer', () => {
    const props = { items, selectedIndex: 1 };
    const { lastFrame } = render(<ControlledSelectInput {...props} />);
    expect(lastFrame()).toMatchSnapshot();
  });
});
