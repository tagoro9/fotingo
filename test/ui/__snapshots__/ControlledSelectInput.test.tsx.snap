// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlledSelectInput /> renders the list of items 1`] = `
"  Item 1
  Item 2"
`;

exports[`<ControlledSelectInput /> renders the selected item with a pointer 1`] = `
"  Item 1
‚ùØ Item 2"
`;
